### ISSUE TITLE
Pipeline corrupted

### ISSUE THREAD

**Issue by jindrichmynarz** [linkedpipes/etl#375]
LP-ETL corrupted a pipeline so that it cannot be executed. While the pipeline's JSON-LD is syntactically valid (tested with Apache Jena's `riot`), executing the pipeline throws this error to the execution monitor:

```
Caused by: org.eclipse.rdf4j.rio.RDFParseException: Could not parse JSONLD
	at org.eclipse.rdf4j.rio.jsonld.JSONLDParser.parse(JSONLDParser.java:74)
	at com.linkedpipes.etl.executor.monitor.execution.PipelineLoader.loadStream(PipelineLoader.java:200)
	... 38 common frames omitted
Caused by: com.fasterxml.jackson.core.JsonParseException: Unexpected character ('<' (code 60)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')
 at [Source: java.io.BufferedReader@6ed527e2; line: 1, column: 2]
```

The corrupted pipeline is [here](https://github.com/linkedpipes/etl/files/966667/lau.jsonld.zip).

**Comment by jindrichmynarz** [comment]
This happened again. In the edit view of the corrupted pipeline it seems that components are duplicated, because they need to be deleted twice. Once for the component connections, once for the component itself to disappear.

**Comment by skodapetr** [comment]
I'm not sure about the duplicated components, the execution should not change the pipeline.

The attached pipeline contains a cycle (Generate LAU1 IRIs -> Add language tags ) which cause failure - a bug is that this problem is not properly communicated with user.

The cycle also prevents frontend from removing the connection - as it tries to iterate over all predecessors and ends up in infinite cycle. So again this is problem with cycle handling but this time in frontend.

**Comment by jindrichmynarz** [comment]
This is also a problem of the back-end that corrupts the pipeline by creating the cycle.

**Comment by skodapetr** [comment]
That is another issue, the backend should not modify the pipeline.  You probably do not have steps to reproduce?

**Comment by jindrichmynarz** [comment]
I can recreate the corrupted pipeline. Then, the pipeline get corrupted *sometimes* when it is executed. I'll try to find if there's a more specific way to corrupt the pipeline.

**Comment by skodapetr** [comment]
And the steps to "recreate" , is it from the json file, or by creating cycle .. as once there is a cycle there  it can starts to snowball.

**Comment by jindrichmynarz** [comment]
So it seems that front-end causes this problem. When I manually recreate the pipeline and try to run it, the executor refuses to start its execution due to an error. When I reopen the pipeline the error is apparent, as there are cycles created between "Remove empty rows" and "Remove empty values" and then "Link NUTS3 2010" and "Map SKOS":

<img width="648" alt="screen shot 2017-05-02 at 12 24 46" src="https://cloud.githubusercontent.com/assets/198642/25614539/a0a9cffa-2f32-11e7-852a-8d27dacda1e8.png">

When I delete the superfluous connections, save, and reopen the pipeline, the connections are still there. Either they are not deleted, or the change is not saved (perhaps due to the parser error), or the connections are recreated by the front-end for some reason.

**Comment by skodapetr** [comment]
Once the cycle is created the frontend can not delete it - that is the aforementioned problem with cycle handling in frontend.

And you recreate the pipeline with or without the cycle?

**Comment by jindrichmynarz** [comment]
I recreated the pipeline without the cycles. They must be introduced by the front-end (as the pipeline is saved?).

**Comment by skodapetr** [comment]
And could you also post the pipeline without cycles? Ie. the one you save and then the cycles appear.

**Comment by jindrichmynarz** [comment]
I can try recreating the pipeline, exporting it at each step, and see when the cycles are introduced.

**Comment by jindrichmynarz** [comment]
Got it. A cycle is introduced after adding another SPARQL Update component to the pipeline. Here's the [pipeline before](https://github.com/linkedpipes/etl/files/970216/lau_before.jsonld.zip) adding the component. Here's the [pipeline after](https://github.com/linkedpipes/etl/files/970215/lau_after.jsonld.zip) adding the component with the cycle in.

However, I'm unsure why this happens at this point, since there are already multiple SPARQL Update components. It seems that when there are multiple instances of the same component, there's is a chance that the front-end will add a cycle to them.

**Comment by skodapetr** [comment]
The component ID is assigned based on the IR provided by JointJS which should be unique - ie. the component type should not play any role. 

So the steps to reproduce are:
- load "pipeline before"
- add SPARQL update
- connect "SPARQL update Cast strings to booleans"  to new SPARQL  update
- save pipeline
- close and open pipeline
and this should create the cycle.

**Comment by jindrichmynarz** [comment]
Corrupted pipeline at the demo instance is [here](https://demo.etl.linkedpipes.com/#/pipelines/edit/canvas?pipeline=https:%2F%2Fdemo.etl.linkedpipes.com%2Fresources%2Fpipelines%2Fcreated-1494315232556).

This bug seems a bit non-deterministic, but it's easy to reproduce.

1. Import "pipeline before"
2. Add SPARQL Update linked from the last instance of the SPARQL Update component in the pipeline
3. Save
4. Close
5. Open
6. Repeat steps 2.-5. until cycles appear.

**Comment by jakubklimek** [comment]
I did this twice and no luck. Maybe a screencast?

**Comment by jindrichmynarz** [comment]
Fixed in 1119fce83501aeee0ca73c466d89aa09ac042433.
